README.md
<img src="Hyperdrive.png" width=51 height=49 alt="Hyperdrive Logo" />

# Hyperdrive [![Build Status](https://travis-ci.org/the-hypermedia-project/Hyperdrive.svg?branch=master)](https://travis-ci.org/the-hypermedia-project/Hyperdrive)

Hyperdrive is a generic Hypermedia API client in Swift. Hyperdrive allows
you to build an application which can evolve at run-time and does not
hard-code implementation details such as URIs and HTTP methods into your
application. You simply enter your API using the root URI and explore it's
funcitonality at run-time by understanding the semantics of the domain
instead of knowledge about the implementation.

## Usage

Below is a short example of using Hyperdrive to communicate with a Hypermedia API.
An API that offers information about how it works at run-time using hyperlinks.

We're going to connect to a [Polls API](https://github.com/apiaryio/polls-api),
an API which allows you to view questions, vote for them and create new
questions with multiple-choice answers.

```swift
let hyperdrive = Hyperdrive()
```

To get started, we will enter the API from its root URI. We will pass it an
asynchronous function to be executed with a result from the API.

```swift
hyperdrive.enter("https://polls.apiblueprint.org/") { result in
  switch result {
    case .Success(let representor):
      print("The API has offered us the following transitions: \(representor.transitions)")

    case .Failure(let error):
      print("Unfortunately there was an error: \(error)")
  }
}
```

On success, we have a Representor, this is a structure representing the API
resource. It includes relations to other resources along with information
about how we can transition from the current state to another.

Our client understands the semantics behind “questions” and explicitly
looks for a transition to them on our API.

```swift
if let questions = representor.transitions["questions"] {
  print("Our API has a transition to a questions resource.")
} else {
  print("Looks like this API doesn’t support questions, or " +
          "the feature was removed.")
}
```

Since our API has a transition to a collection of questions, let’s retrieve
them and take a look:

```swift
hyperdrive.request(questions) { result in
  switch result {
    case .Success(let representor):
      print("We’ve received a representor for the questions.")

    case .Failure(let error):
      print("There was a problem retreiving the questions: \(error)")
  }
}
```

On success, we have another representor representing the questions resource in
our API.

```swift
if let questions = representor.representors["questions"] {
  // Our representor includes a collection of Question resources.
  // Let’s use map to call viewQuestion for each one
  map(questions, viewQuestion)
} else {
  print("Looks like there are no questions yet.")
  print("Perhaps the API offers us the ability to create a question?")
}
```

With every question in this resource, we will call our `viewQuestion` function:

```swift
func viewQuestion(question:Representor<HTTPTransition>) {
  print(question.attributes["question"])

  if let choices = question.representors["choices"] {
    for choice in choices {
      let text = choice.attributes["choice"]
      let votes = choice.attributes["votes"]
      print('-> \(text) (\(votes))')
    }
  } else {
    print("-> This question does not have any choices.")
  }
}
```

### Transitioning from one state to another

A representor includes information on how to transition from one state to
another. For example, we might be presented with the ability to delete
one of our questions:

```swift
if let delete = question.transitions["delete"] {
  // We can perform this transition with Hyperdrive
  hyperdrive.request(delete)
} else {
  print("The API doesn’t offer us the ability to delete a question.")
  print("Let’s gracefully handle the lack of this feature and " +
          "remove deletion from our user interface.")
}
```

We may also be presented with an option to vote on a choice:

```swift
if let vote = choice.transitions["vote"] {
  hyperdrive.request(vote)
}
```

#### Transitions with attributes

A transition may also provide attributes for performing the transition. For
example, our API may afford us to perform a transition to create a new
question.

```swift
if let create = questions.transitions["create"] {
  let attributes = [
    "question": "Favourite programming language?",
    "choices": [
      "Swift",
      "Python",
      "Ruby",
    ]
  ]

  hyperdrive.request(create, attributes)
} else {
  // Our API doesn’t allow us to create a question. We should remove the
  // ability to create a question in our user interface.
  // This transition may only be available to certain users.
}
```

Transitions also include the available attributes you can send for run-time
introspection.

```swift
create.attributes
```

This allows you to generate user interface that can adapt to changes from the
API. You can also use validation for an attribute.

## Content Types

Hyperdrive supports the following Hypermedia content-types:

- [Siren](https://github.com/kevinswiber/siren) (`application/vnd.siren+json`)
- [HAL](http://stateless.co/hal_specification.html) (`application/hal+json`)

For APIs which do not support Hypermedia content types, you may use an API
description in form of an [API Blueprint](Blueprint.md) to load these controls.

## Installation

[CocoaPods](http://cocoapods.org/) is the recommended installation method.

```ruby
pod 'Hyperdrive'
```

## Elsehwere

- [RxHyperdrive](https://github.com/kylef/RxHyperdrive) - RxSwift extensions for using Hyperdrive.
- [Polls Application](https://github.com/apiaryio/polls-app) - An example iOS application talking to the Polls Hypermedia API.
- [Starship](https://github.com/kylef/Starship) - A generic API client application for iOS using Hyperdrive.
- [Mockingdrive](http://github.com/kylef/Mockingdrive) - Tool for easily stubbing Hypermedia HTTP requests in tests.

## License

Hyperdrive is released under the MIT license. See [LICENSE](LICENSE).



CHANGELOG.md
# Hyperdrive Changelog
## 0.3.0
### Bug Fixes

- HyperBlueprint now supports the latest version of API Blueprint AST, which
  does not contain MSON AST.


Hyperdrive/Hyperdrive.swift
//
//  Hyperdrive.swift
//  Hyperdrive
//
//  Created by Kyle Fuller on 08/04/2015.
//  Copyright (c) 2015 Apiary. All rights reserved.
//

import Foundation
import Representor
import URITemplate
import Result


/// Map a dictionaries values
func map<K,V>(source:[K:V], transform:(V -> V)) -> [K:V] {
  var result = [K:V]()

  for (key, value) in source {
    result[key] = transform(value)
  }

  return result
}

/// Returns an absolute URI for a URI given a base URL
func absoluteURI(baseURL: NSURL?) -> (uri: String) -> String {
  return { uri in
    return NSURL(string: uri, relativeToURL: baseURL)?.absoluteString ?? uri
  }
}

/// Traverses a representor and ensures that all URIs are absolute given a base URL
func absoluteRepresentor(baseURL: NSURL?) -> (original: Representor<HTTPTransition>) -> Representor<HTTPTransition> {
  return { original in
    let transitions = map(original.transitions) { transition in
      return HTTPTransition(uri: absoluteURI(baseURL)(uri: transition.uri)) { builder in
        builder.method = transition.method
        builder.suggestedContentTypes = transition.suggestedContentTypes

        for (name, attribute) in transition.attributes {
          builder.addAttribute(name, value: attribute.value, defaultValue: attribute.defaultValue)
        }

        for (name, parameter) in transition.parameters {
          builder.addParameter(name, value: parameter.value, defaultValue: parameter.defaultValue)
        }
      }
    }

    let representors = map(original.representors) { representors in
      representors.map(absoluteRepresentor(baseURL))
    }

    return Representor(transitions: transitions, representors: representors, attributes: original.attributes, metadata: original.metadata)
  }
}


public typealias RepresentorResult = Result<Representor<HTTPTransition>, NSError>
public typealias RequestResult = Result<NSMutableURLRequest, NSError>
public typealias ResponseResult = Result<NSHTTPURLResponse, NSError>


/// A hypermedia API client
public class Hyperdrive {
  public static var errorDomain:String {
    return "Hyperdrive"
  }

  private let session:NSURLSession

  /// An array of the supported content types in order of preference
  let preferredContentTypes:[String]

  /** Initialize hyperdrive
  - parameter preferredContentTypes: An optional array of the supported content types in order of preference, when this is nil. All types supported by the Representor will be used.
  */
  public init(preferredContentTypes:[String]? = nil) {
    let configuration = NSURLSessionConfiguration.defaultSessionConfiguration()
    session = NSURLSession(configuration: configuration)
    self.preferredContentTypes = preferredContentTypes ?? HTTPDeserialization.preferredContentTypes
  }

  // MARK: -

  /// Enter a hypermedia API given the root URI
  public func enter(uri:String, completion:(RepresentorResult -> Void)) {
    request(uri, completion:completion)
  }

  // MARK: Subclass hooks

  /// Construct a request from a URI and parameters
  public func constructRequest(uri:String, parameters:[String:AnyObject]? = nil) -> RequestResult {
    let expandedURI = URITemplate(template: uri).expand(parameters ?? [:])

    let error = NSError(domain: Hyperdrive.errorDomain, code: 0, userInfo: [NSLocalizedDescriptionKey: "Creating NSURL from given URI failed"])
    return Result(NSURL(string: expandedURI), failWith: error).map { URL in
      let request = NSMutableURLRequest(URL: URL)
      request.setValue(preferredContentTypes.joinWithSeparator(", "), forHTTPHeaderField: "Accept")
      return request
    }
  }

  public func constructRequest(transition:HTTPTransition, parameters:[String:AnyObject]?  = nil, attributes:[String:AnyObject]? = nil) -> RequestResult {
    return constructRequest(transition.uri, parameters:parameters).map { request in
      request.HTTPMethod = transition.method

      if let attributes = attributes {
        request.HTTPBody = self.encodeAttributes(attributes, suggestedContentTypes: transition.suggestedContentTypes)
      }

      return request
    }
  }

  func encodeAttributes(attributes:[String:AnyObject], suggestedContentTypes:[String]) -> NSData? {
    let JSONEncoder = { (attributes:[String:AnyObject]) -> NSData? in
      return try? NSJSONSerialization.dataWithJSONObject(attributes, options: NSJSONWritingOptions(rawValue: 0))
    }

    let encoders:[String:([String:AnyObject] -> NSData?)] = [
      "application/json": JSONEncoder
    ]

    for contentType in suggestedContentTypes {
      if let encoder = encoders[contentType] {
        return encoder(attributes)
      }
    }

    return JSONEncoder(attributes)
  }

  public func constructResponse(request:NSURLRequest, response:NSHTTPURLResponse, body:NSData?) -> Representor<HTTPTransition>? {
    if let body = body {
      let representor = HTTPDeserialization.deserialize(response, body: body)
      if let representor = representor {
        return absoluteRepresentor(response.URL)(original: representor)
      }
    }

    return nil
  }

  // MARK: Perform requests

  func request(request:NSURLRequest, completion:(RepresentorResult -> Void)) {
    let dataTask = session.dataTaskWithRequest(request, completionHandler: { (body, response, error) -> Void in
      if let error = error {
        dispatch_async(dispatch_get_main_queue()) {
          completion(.Failure(error))
        }
      } else {
        let representor = self.constructResponse(request, response:response as! NSHTTPURLResponse, body: body) ?? Representor<HTTPTransition>()
        dispatch_async(dispatch_get_main_queue()) {
          completion(.Success(representor))
        }
      }
    })

    dataTask.resume()
  }

  /// Perform a request with a given URI and parameters
  public func request(uri:String, parameters:[String:AnyObject]? = nil, completion:(RepresentorResult -> Void)) {
    switch constructRequest(uri, parameters: parameters) {
    case .Success(let request):
      self.request(request, completion:completion)
    case .Failure(let error):
      completion(.Failure(error))
    }
  }

  /// Perform a transition with a given parameters and attributes
  public func request(transition:HTTPTransition, parameters:[String:AnyObject]? = nil, attributes:[String:AnyObject]? = nil, completion:(RepresentorResult -> Void)) {
    let result = constructRequest(transition, parameters: parameters, attributes: attributes)

    switch result {
    case .Success(let request):
      self.request(request, completion:completion)
    case .Failure(let error):
      completion(.Failure(error))
    }
  }
}


HyperdriveTests/HyperdriveTests.swift
//
//  HyperdriveTests.swift
//  HyperdriveTests
//
//  Created by Kyle Fuller on 08/04/2015.
//  Copyright (c) 2015 Apiary. All rights reserved.
//

import Foundation
import XCTest
import Representor
import Hyperdrive


class HyperdriveTests: XCTestCase {
  var hyperdrive:Hyperdrive!

  override func setUp() {
    hyperdrive = Hyperdrive()
  }

  // MARK: Constructing a request from a URI

  func testConstructingRequestReturnsWithAcceptHeaderRequest() {
    let result = hyperdrive.constructRequest("https://hyperdrive-tests.fuller.li/")

    switch result {
    case .Success(let request):
      let header = request.allHTTPHeaderFields!["Accept"]
      XCTAssertEqual(header, "application/vnd.siren+json, application/hal+json")
    case .Failure(let error):
      XCTFail(error.description)
    }
  }

  func testConstructingRequestReturnsWithAcceptHeaderRequestUsingSuppliedContentTypes() {
    hyperdrive = Hyperdrive(preferredContentTypes: ["application/hal+json"])
    let result = hyperdrive.constructRequest("https://hyperdrive-tests.fuller.li/")

    switch result {
    case .Success(let request):
      let header = request.allHTTPHeaderFields!["Accept"]
      XCTAssertEqual(header, "application/hal+json")
    case .Failure(let error):
      XCTFail(error.description)
    }
  }

  func testConstructingRequestExpandsURITemplate() {
    let result = hyperdrive.constructRequest("https://hyperdrive-tests.fuller.li/{username}", parameters:["username": "kyle"])

    switch result {
    case .Success(let request):
      XCTAssertEqual(request.URL?.absoluteString, "https://hyperdrive-tests.fuller.li/kyle")
    case .Failure(let error):
      XCTFail(error.description)
    }
  }

  // MARK: Constructing a request from a transition

  func testConstructingRequestFromTransitionReturnsWithAcceptHeaderRequest() {
    let transition = HTTPTransition(uri: "https://hyperdrive-tests.fuller.li/users") { builder in }
    let result = hyperdrive.constructRequest(transition)

    switch result {
    case .Success(let request):
      let header = request.allHTTPHeaderFields!["Accept"]
      XCTAssertEqual(header, "application/vnd.siren+json, application/hal+json")
    case .Failure(let error):
      XCTFail(error.description)
    }
  }

  func testConstructingRequestFromTransitionExpandsURITemplate() {
    let transition = HTTPTransition(uri: "https://hyperdrive-tests.fuller.li/{username}") { builder in }
    let result = hyperdrive.constructRequest(transition, parameters:["username": "kyle"])

    switch result {
    case .Success(let request):
      XCTAssertEqual(request.URL?.absoluteString, "https://hyperdrive-tests.fuller.li/kyle")
    case .Failure(let error):
      XCTFail(error.description)
    }
  }

  func testConstructingRequestFromTransitionWithMethod() {
    let transition = HTTPTransition(uri: "https://hyperdrive-tests.fuller.li/kyle") { builder in
      builder.method = "PATCH"
    }
    let result = hyperdrive.constructRequest(transition)

    switch result {
    case .Success(let request):
      XCTAssertEqual(request.HTTPMethod, "PATCH")
    case .Failure(let error):
      XCTFail(error.description)
    }
  }

  func testConstructingRequestFromTransitionWithJSONAttributes() {
    let transition = HTTPTransition(uri: "https://hyperdrive-tests.fuller.li/users") { builder in
      builder.suggestedContentTypes = ["application/json"]
    }
    let result = hyperdrive.constructRequest(transition, attributes:["username": "kyle"])

    switch result {
    case .Success(let request):
      let body = request.HTTPBody!
      let decodedBody = try? NSJSONSerialization.JSONObjectWithData(body, options: NSJSONReadingOptions(rawValue: 0)) as! NSDictionary
      XCTAssertEqual(decodedBody, ["username": "kyle"])
    case .Failure(let error):
      XCTFail(error.description)
    }
  }

  func xtestConstructingRequestFromTransitionWithFormEncodedAttributes() {

  }

  // MARK: Constructing a response

  func testConstructingResponseFromSirenBody() {
    let URL = NSURL(string: "https://hyperdrive-tests.fuller.li/users")!
    let request = NSURLRequest(URL: URL)
    let response = NSHTTPURLResponse(URL: URL, statusCode: 200, HTTPVersion: "1.1", headerFields: ["Content-Type": "application/vnd.siren+json"])!
    let attributes = [
      "properties": [
        "test": "hello world"
      ]
    ]
    let body = try! NSJSONSerialization.dataWithJSONObject(attributes, options: NSJSONWritingOptions(rawValue: 0))

    let representor = hyperdrive.constructResponse(request, response: response, body:body)!
    XCTAssertEqual(representor.attributes["test"] as? String, "hello world")
  }
}



HyperdriveTests/HyperdriveGitHubTest.swift
//
//  HyperdriveGitHubTest.swift
//  Hyperdrive
//
//  Created by Z on 8/8/15.
//  Copyright (c) 2015 Apiary. All rights reserved.
//

import Foundation
import XCTest
import Representor
import Hyperdrive

class GitHubAdapterTests: XCTestCase {

  func testGitHubRoot() {
    let hyperdrive = Hyperdrive(preferredContentTypes: ["application/vnd.github.v3+json"])
    
    let expectation = expectationWithDescription("...")
    
    hyperdrive.enter("https://api.github.com/") { result in
      switch result {
      case .Success(let representor):
        println("The API has offered us the following transitions: \(representor.transitions)")
        
      case .Failure(let error):
        println("Unfortunately there was an error: \(error)")
      }
      
      expectation.fulfill()
    }
    
    waitForExpectationsWithTimeout(10) { error in
      if let error = error {
        print("Error: \(error.localizedDescription)")
      }
    }
  }
  
}

HyperdriveTests/HyperBlueprintTests.swift
//
//  HyperBlueprintTests.swift
//  Hyperdrive
//
//  Created by Kyle Fuller on 12/04/2015.
//  Copyright (c) 2015 Apiary. All rights reserved.
//

import Foundation
import XCTest
import Representor
import Hyperdrive


class HyperBlueprintTests: XCTestCase {
  var blueprint:Blueprint!
  var hyperdrive:HyperBlueprint!

  override func setUp() {
    let listAction = Action(name: "List", description: nil, method: "GET", parameters: [], uriTemplate: nil, relation: "questions", examples: nil)
    let createAction = Action(name: "Create", description: nil, method: "POST", parameters: [], uriTemplate: nil, relation: "create", examples: nil)
    let listResource = Resource(name: "Questions", description: nil, uriTemplate: "/questions", parameters: [], actions: [createAction, listAction])

    let viewAction = Action(name: "View", description: nil, method: "GET", parameters: [], uriTemplate: nil, relation: "question", examples: nil)
    let viewResource = Resource(name: "Detail", description: nil, uriTemplate: "/questions/{id}", parameters: [], actions: [viewAction])

    let resourceGroup = ResourceGroup(name: "Questions", description: nil, resources: [listResource, viewResource])
    blueprint = Blueprint(name: "Polls", description: nil, resourceGroups: [resourceGroup])
    hyperdrive = HyperBlueprint(blueprint: blueprint, baseURL: NSURL(string: "https://polls.apiblueprint.org/")!)
  }

  // MARK: Root Resource

  func testRootResourceIncludesGETAction() {
    let representor = hyperdrive.rootRepresentor()
    XCTAssertEqual(representor.transitions["questions"]?.first?.uri, "https://polls.apiblueprint.org/questions")
  }

  // MARK:

  func testResourceForResponse() {
    let URL = NSURL(string: "https://polls.apiblueprint.org/questions/5")!
    let response = NSHTTPURLResponse(URL: URL, statusCode: 200, HTTPVersion: nil, headerFields: nil)!
    let resource = hyperdrive.resourceForResponse(response)!

    XCTAssertEqual(resource.name, "Detail")
    XCTAssertEqual(resource.uriTemplate, "/questions/{id}")
  }

  // MARK: Constructing a response

  func testConstructingResponseShowsJSONAttributes() {
    let attributes = ["question": "Favourite Programming Language?"]
    let URL = NSURL(string: "https://polls.apiblueprint.org/questions/5")!
    let request = NSURLRequest(URL: URL)
    let response = NSHTTPURLResponse(URL: URL, statusCode: 200, HTTPVersion: nil, headerFields: ["Content-Type": "application/json"])!
    let body = try! NSJSONSerialization.dataWithJSONObject(attributes, options: NSJSONWritingOptions(rawValue: 0))

    let representor = hyperdrive.constructResponse(request, response: response, body: body)!

    XCTAssertEqual(representor.attributes as NSDictionary, attributes)
  }

  func testConstructingResponseShowsTransitions() {
    let URL = NSURL(string: "https://polls.apiblueprint.org/questions")!
    let request = NSURLRequest(URL: URL)
    let response = NSHTTPURLResponse(URL: URL, statusCode: 200, HTTPVersion: nil, headerFields: ["Content-Type": "application/json"])!
    let body = try! NSJSONSerialization.dataWithJSONObject([], options: NSJSONWritingOptions(rawValue: 0))

    let representor = hyperdrive.constructResponse(request, response: response, body: body)!
    let createTransition = representor.transitions["create"]?.first

    XCTAssertTrue(createTransition != nil)
    XCTAssertEqual(createTransition?.uri, "https://polls.apiblueprint.org/questions")
    XCTAssertEqual(createTransition?.method, "POST")
  }

  func testConstructingResponseIncludesSelfTransition() {
    let URL = NSURL(string: "https://polls.apiblueprint.org/questions")!
    let request = NSURLRequest(URL: URL)
    let response = NSHTTPURLResponse(URL: URL, statusCode: 200, HTTPVersion: nil, headerFields: ["Content-Type": "application/json"])!
    let body = try! NSJSONSerialization.dataWithJSONObject([], options: NSJSONWritingOptions(rawValue: 0))

    let representor = hyperdrive.constructResponse(request, response: response, body: body)!
    let transition = representor.transitions["self"]?.first

    XCTAssertNotNil(transition)
    XCTAssertEqual(transition?.uri, "https://polls.apiblueprint.org/questions")
    XCTAssertEqual(transition?.method, "GET")
  }

  func testConstructingResponseHidesTransitionsNotIncludedInAllowHeader() {
    let URL = NSURL(string: "https://polls.apiblueprint.org/questions")!
    let request = NSURLRequest(URL: URL)
    let headers = [
      "Allow": "HEAD, GET",
      "Content-Type": "application/json",
    ]
    let response = NSHTTPURLResponse(URL: URL, statusCode: 200, HTTPVersion: nil, headerFields: headers)!
    let body = try! NSJSONSerialization.dataWithJSONObject([], options: NSJSONWritingOptions(rawValue: 0))

    let representor = hyperdrive.constructResponse(request, response: response, body: body)!
    let createTransition = representor.transitions["create"]

    XCTAssertNil(createTransition)
  }

  func testConstructingResponseShowsWebLinkingHeaders() {
    let attributes = ["question": "Favourite Programming Language?"]
    let URL = NSURL(string: "https://polls.apiblueprint.org/questions/5")!
    let request = NSURLRequest(URL: URL)
    let headers = [
      "Content-Type": "application/json",
      "Link": "<https://polls.apiblueprint.org/questions/6>; rel=\"next\", <https://polls.apiblueprint.org/questions/4>; rel=\"prev\"; type=\"foo\"",
    ]
    let response = NSHTTPURLResponse(URL: URL, statusCode: 200, HTTPVersion: nil, headerFields: headers)!
    let body = try! NSJSONSerialization.dataWithJSONObject(attributes, options: NSJSONWritingOptions(rawValue: 0))

    let representor = hyperdrive.constructResponse(request, response: response, body: body)!
    let nextTransition = representor.transitions["next"]?.first
    let prevTransition = representor.transitions["prev"]?.first

    XCTAssertEqual(prevTransition?.uri, "https://polls.apiblueprint.org/questions/4")
    XCTAssertEqual(prevTransition?.method, "GET")
    XCTAssertEqual(prevTransition!.suggestedContentTypes, ["foo"])
    XCTAssertEqual(nextTransition?.uri, "https://polls.apiblueprint.org/questions/6")
    XCTAssertEqual(nextTransition?.method, "GET")
    XCTAssertEqual(nextTransition!.suggestedContentTypes, [])
  }
}
